gatsby-node.js

```diff
@@ -35,27 +42,102 @@ exports.createPages = async ({ actions, graphql, reporter }) => {
   const postOrPage = result.data.allMarkdownRemark.edges
     .filter(edge => edge.node.frontmatter.template !== "announcement");
 
+  // Render pages based on their frontmatter template, if available, or a
+  // default template at src/templates/default.js
   postOrPage.forEach(({ node }) => {
-    debugger;
     const templateName = node.frontmatter.template || `default`
     const template = path.resolve(path.join('src/templates/', `${templateName}.js`))
     createPage({
       path: node.fields.slug,
       component: template,
-      context: { id: node.id }, // additional data can be passed via context
+      // additional data can be passed via context
+      context: { id: node.id, locale: node.fields.locale },
     })
   })
 }
 
-exports.onCreateNode = ({ node, actions, getNode }) => {
+function langFromFilename(filename) {
+  const potentialMatch = filename.match(/.*\.(\w+)\.\w+$/i)
+  return potentialMatch && potentialMatch[1] || null
+}
+
+function localePath(path, lang) {
+  let pathComponents = path.split('/').filter(s => s !== '')
+  let last = pathComponents.pop()
+
+  if (last === `index.${lang}`) {
+    last = ''
+  } else {
+    last = last.replace(new RegExp(`\\.${lang}$`, 'i'), '')
+  }
+
+  pathComponents.push(last)
+
+  return `/${lang}/${pathComponents.join('/')}`
+}
+
+exports.onCreateNode = async ({ graphql, node, actions, getNode }) => {
   const { createNodeField } = actions
 
+  // set a slug for all markdown nodes, and use any supported 639-1 language
+  // code prepended to the file extension to set a language-specific URL root
   if (node.internal.type === `MarkdownRemark`) {
-    const value = createFilePath({ node, getNode })
+    let lang = langFromFilename(node.fileAbsolutePath)
+    let slug = createFilePath({ node, getNode })
+
+    if (lang && supportedLocales.includes(lang)) {
+      if (lang !== defaultLocale) {
+        slug = languagePath(slug, lang)
+      } else {
+        lang = null
```

Comment:
Might be not reading the code carefully enough, but do we need this else case here? Seems like this 
```
let lang = langFromFilename(node.fileAbsolutePath)
```
 already defines `lang` as null if there was no match?

Reply:
Took me a minute- this still is handy because of the supported lang list, but I'm going to clean up the control flow